lib/ex_rlp/decode.ex

```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
 
-  # When we don't have any RLP left, return the accumulator
-  defp decode_item(<<>>, result) do
+  defp decode_item("", result) when is_list(result) do
     Enum.reverse(result)
   end
 
-  # Decodes the head represents an item to be added directly
-  # to the result.
+  defp decode_item("", result) do
+    result
+  end
+
   defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix < 128 do
-    decode_item(tail, [<<prefix>> | result])
+    new_item = <<prefix>>
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
   end
 
-  # Decodes medium length-binary?
   defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 183 do
     {new_item, new_tail} = decode_medium_binary(prefix, tail, 128)
 
-    decode_item(new_tail, [new_item | result])
+    new_result = add_new_item(result, new_item)
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes long length-binary?
   defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix < 192 do
     {new_item, new_tail} = decode_long_binary(be_size_prefix, tail, 183)
 
-    decode_item(new_tail, [new_item | result])
+    new_result = add_new_item(result, new_item)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix == 192 do
+    new_item = []
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
+  end
+
+  defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 247 do
+    {list, new_tail} = decode_medium_binary(prefix, tail, 192)
+
+    new_result = add_decoded_list(result, list)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) do
+    {list, new_tail} = decode_long_binary(be_size_prefix, tail, 247)
+
+    new_result = result |> add_decoded_list(list)
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes an empty list
-  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result)
-       when be_size_prefix == 192 do
-    decode_item(tail, [[] | result])
+  @spec add_new_item(ExRLP.t(), ExRLP.t()) :: ExRLP.t()
+  def add_new_item(nil, new_item) do
+    new_item
   end
 
-  defp decode_item(<<(<<prefix>>), tail::binary>>, result) do
-    {list_bin, new_tail} =
-      if prefix <= 247 do
-        decode_medium_binary(prefix, tail, 192)
-      else
-        decode_long_binary(prefix, tail, 247)
-      end
+  def add_new_item(result, new_item) do
+    [new_item | result]
+  end
+
+  @spec add_decoded_list(ExRLP.t(), binary()) :: ExRLP.t()
+  defp add_decoded_list(nil, rlp_list_binary) do
+    rlp_list_binary
+    |> decode_item([])
+    |> Enum.reverse()
+  end
 
-    next_list = decode_item(list_bin, [])
+  defp add_decoded_list(result, rlp_list_binary) do
+    list_items = decode_item(rlp_list_binary, [])
 
-    decode_item(new_tail, [next_list | result])
```

Comment:
This does remove a tail-call optimization, right?

Reply:
no, it's moved `add_decoded_list` function